--- gnome-icon-theme-symbolic-3.8.0.1-orig/Makefile.in	2013-03-27 06:55:41.000000000 +0800
+++ gnome-icon-theme-symbolic-3.8.0.1/Makefile.in	2013-04-05 23:18:23.000000000 +0800
@@ -790,7 +790,7 @@
 
 install-data-hook:
 	if test -z "$(DESTDIR)" ; then \
-		$(GTK_UPDATE_ICON_CACHE) -q $(DESTDIR)$(themedir); \
+		$(GTK_UPDATE_ICON_CACHE) -qt $(DESTDIR)$(themedir); \
 	fi
 
 -include $(top_srcdir)/git.mk
