--- yelp-build	2013-05-11 23:19:38.000000000 +0800
+++ /Users/asbel/gnote/10.7/inst/bin/yelp-build	2013-05-11 23:29:31.000000000 +0800
@@ -346,10 +346,10 @@
         html_cache_file=`(cd $(dirname "$html_cache_file") && pwd)`/`basename "$html_cache_file"`
     else
         html_cache_file_is_tmp="yes"
-        html_cache_file=`mktemp`
+        html_cache_file=`mktemp -t yelp`
         yelp_cache -o "$html_cache_file" "$@"
     fi
-    html_tmp_infile=`mktemp`
+    html_tmp_infile=`mktemp -t yelp`
     yelp_cache_in "$@" > "$html_tmp_infile"
     (
         echo '<xsl:stylesheet'
@@ -564,11 +564,11 @@
         epub_cache_file=`(cd $(dirname "$epub_cache_file") && pwd)`/`basename "$epub_cache_file"`
     else
         epub_cache_file_is_tmp="yes"
-        epub_cache_file=`mktemp`
+        epub_cache_file=`mktemp -t yelp`
         yelp_cache -o "$epub_cache_file" "$@"
     fi
     html_cache_file="$epub_cache_file"
-    epub_data_out=`mktemp -d`
+    epub_data_out=`mktemp -t yelp -d`
     html_out="$epub_data_out/OPS"
     mkdir "$html_out"
     mkdir "$html_out/yelp"
